src/include/migraphx/op/nonmaxsuppression.hpp

```diff
@@ -128,100 +127,95 @@ struct nonmaxsuppression
-            const float* batch_boxes = boxes + bidx * box_num * 4;
-            std::priority_queue<std::pair<float, int64_t>> sorted_boxes;
-            auto insert_to_sorted_boxes =
-                make_function_output_iterator([&](const auto& x) { sorted_boxes.push(x); });
-
-            int64_t box_idx = 0;
-            transform_if(
-                scores.begin() + score_offset,
-                scores.begin() + score_offset + box_num,
-                insert_to_sorted_boxes,
-                [&](auto sc) {
-                    box_idx++;
-                    return sc >= score_threshold;
-                },
-                [&](auto sc) { return std::make_pair(sc, box_idx - 1); });
-
-            selected_boxes_inside_class.clear();
-            // Get the next box with top score, filter by iou_threshold
-            while(!sorted_boxes.empty() &&
-                  selected_boxes_inside_class.size() < max_output_boxes_per_class)
-            {
-                const std::pair<float, int64_t>& next_top_score = sorted_boxes.top();
-
-                // Check with existing selected boxes for this class, suppress if exceed the IOU
-                // (Intersection Over Union) threshold
-                bool not_selected = std::any_of(
-                    selected_boxes_inside_class.begin(),
-                    selected_boxes_inside_class.end(),
-                    [&](auto selected_index) {
-                        return this->suppress_by_iou(batch_box(batch_boxes, next_top_score.second),
-                                                     batch_box(batch_boxes, selected_index.second),
-                                                     iou_threshold);
-                    });
-
-                if(not not_selected)
-                {
-                    selected_boxes_inside_class.push_back(next_top_score);
-                    selected_indices.push_back(bidx);
-                    selected_indices.push_back(cidx);
-                    selected_indices.push_back(next_top_score.second);
-                }
-                sorted_boxes.pop();
-            }
-        });
-
-        result.visit([&](auto out) {
-            std::copy(selected_indices.begin(), selected_indices.end(), out.begin());
+        float iou_threshold   = (args.size() > 3) ? (args.at(3).at<float>()) : 0.0f;
+        float score_threshold = (args.size() > 4) ? (args.at(4).at<float>()) : 0.0f;
+
+        result.visit([&](auto output) {
+            visit_all(args[0], args[1])([&](auto boxes, auto scores) {
+                std::fill(output.begin(), output.end(), 0);
+                const auto& lens       = scores.get_shape().lens();
+                const auto num_batches = lens[0];
+                const auto num_classes = lens[1];
+                const auto num_boxes   = boxes.get_shape().lens()[1];
+                // boxes of a class with NMS applied [score, index]
+                std::vector<std::pair<float, int64_t>> selected_boxes_inside_class;
```

Comment:
We usually use double in the ref implementation to get better accuracy, maybe this would help here too.

Reply:
The ONNX spec actually has both the boxes and scores limited to floats. https://github.com/onnx/onnx/blob/main/docs/Operators.md#NonMaxSuppression
I think that's why this operator didn't use the visitor before. I made the change to be consistent with the other operators and in case the spec changes later on.